get_adjacent_post:  adjacent post query plugin compatibility

ramonjd opened the pull request:
Follow up to https://github.com/WordPress/wordpress-develop/pull/10394

Tests out an idea suggested in https://core.trac.wordpress.org/ticket/64390#comment:10 Props @azaozz 

> [!NOTE]
> This is mainly to discuss whether it's worth the change, if so good, if not, I can close.

## Description

This PR is a follow-up to [#10394](https://github.com/WordPress/wordpress-develop/pull/10394) that makes the deterministic ordering changes more defensive and plugin-friendly. It ensures that when plugins modify the `get_{$adjacent}_post_where` or `get_{$adjacent}_post_sort` filters, the deterministic logic is not applied on top of their modifications.

### Problem

In PR #10394, deterministic ID-based ordering was added to fix adjacent post navigation for posts with identical dates. However, the implementation applied the deterministic logic before the filters ran, which meant:

1. Plugins modifying the WHERE or SORT clauses via filters would receive SQL that already included the deterministic changes
2. Plugins couldn't opt out or modify the SQL in ways that might conflict with the deterministic logic
3. This could break plugins that parse or manipulate the SQL clauses

### Solution

This PR implements a defensive approach that:

1. **Applies filters first** with the original (non-deterministic) SQL clauses
2. **Only applies deterministic logic if filters don't modify the clauses** - checks if the filtered value equals the original prepared value
3. **Respects plugin modifications** - when filters modify the SQL, their changes are preserved and deterministic logic is not applied

**Implementation:**
- For WHERE clause: Only applies ID-based fallback if `$where === $where_prepared` (filter didn't modify it)
- For SORT clause: Only applies ID-based sorting if `$sort === $sort_prepared` (filter didn't modify it)



## Test Steps

### Automated Tests

Run the PHPUnit tests:

```bash
npm run test:php -- --filter Tests_Link_GetAdjacentPost
```


### Manual Testing

#### 1. Verify Default Behavior (No Filters)

1. **Create test scenario:**
   - Create multiple posts with identical `post_date` values (bulk publish drafts)
   - Navigate between posts

2. **Expected Results:**
   - Navigation works deterministically, you should be able to navigate through all posts with identical dates

#### 2. Verify Filter Modifications Are Respected

**Test WHERE Filter:**

1. **Add test filter to `functions.php` or a test plugin:**

```php
// Test filter for next post WHERE clause
add_filter( 'get_next_post_where', function( $where ) {
	// Modify the WHERE clause - deterministic fallback should NOT be applied.
	return $where . ' AND 1=1';
}, 10, 1 );

// Test filter for previous post WHERE clause
add_filter( 'get_previous_post_where', function( $where ) {
	// Modify the WHERE clause - deterministic fallback should NOT be applied.
	return $where . ' AND 1=1';
}, 10, 1 );
```

2. **Verify behavior:**
   - Navigate between posts, rhe deterministic ID fallback should NOT be applied (no `AND p.ID` in WHERE clause)

**Test SORT Filter:**

1. **Remove WHERE filters, add SORT filter:**

```php
add_filter( 'get_next_post_sort', function( $sort, $post, $order ) {
	// Remove ID from sort - deterministic ID sort should NOT be applied.
	return "ORDER BY p.post_date $order LIMIT 1";
}, 10, 3 );

add_filter( 'get_previous_post_sort', function( $sort, $post, $order ) {
	return "ORDER BY p.post_date $order LIMIT 1";
}, 10, 3 );
```

2. **Verify behavior:**
   - Create posts with identical dates
   - Navigate between posts, deterministic ID sort is NOT applied

#### 3. Verify Unmodified Filters Still Get Deterministic Logic

1. **Add filters that don't modify the clauses:**

```php
add_filter( 'get_next_post_where', function( $where ) {
	// Return unchanged - deterministic fallback should be applied.
	return $where;
}, 10, 1 );

add_filter( 'get_next_post_sort', function( $sort ) {
	// Return unchanged - deterministic ID sort should be applied.
	return $sort;
}, 10, 1 );
```

2. **Verify behavior:**
   - Navigation works deterministically, you should be able to navigate through all posts with identical dates



Trac ticket: https://core.trac.wordpress.org/ticket/64390

ramonjd committed:
Add deterministic fallback for adjacent post queries only when the query hasn't been modified.

This update introduces a deterministic fallback for the SQL `WHERE` and `ORDER BY` clauses in the `get_adjacent_post` function when posts have identical dates. The fallback is applied only if the respective clauses have not been modified by filters, ensuring consistent behavior.

Unit tests have been added to verify the correct application of this fallback under various conditions, including scenarios where filters are applied or not.

See Trac ticket https://core.trac.wordpress.org/ticket/64390.

ramonjd committed:
formatting

ramonjd commented on src/wp-includes/link-template.php:
```diff
@@ -1814,6 +1814,7 @@ function get_next_post( $in_same_term = false, $excluded_terms = '', $taxonomy =
  * Can either be next or previous post.
  *
  * @since 2.5.0
+ * @since 6.9.1 Adds deterministic fallback for sort clause if not modified by a filter.
```
Wasn't sure whether to also add a 6.9.0 entry for historical purposes

ramonjd commented on src/wp-includes/link-template.php:
```diff
@@ -1980,7 +1981,6 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 	 *
 	 * @since 2.5.0
 	 * @since 4.4.0 Added the `$taxonomy` and `$post` parameters.
-	 * @since 6.9.0 Adds ID-based fallback for posts with identical dates in adjacent post queries.
```
This okay to remove, or should we have some ref to historical changes?

ramonjd committed:
formatting comment

peterwilsoncc commented on tests/phpunit/tests/link/getAdjacentPost.php:
```diff
@@ -622,6 +622,236 @@ public function test_get_adjacent_post_with_identical_dates() {
+		$this->assertInstanceOf( 'WP_Post', $next );
+		$this->assertEquals( $post_ids[3], $next->ID );
+
+		remove_all_filters( 'get_next_post_where' );
```
Test suite does this automatically so it can be removed. 

:1234: Applies in the tests below so I won't repeat myself.


```suggestion
```

peterwilsoncc commented on tests/phpunit/tests/link/getAdjacentPost.php:
```diff
@@ -622,6 +622,236 @@ public function test_get_adjacent_post_with_identical_dates() {
+		// Next post should be the 4th post (higher ID, same date) - deterministic.
+		$next = get_adjacent_post( false, '', false );
+		$this->assertInstanceOf( 'WP_Post', $next );
+		$this->assertEquals( $post_ids[3], $next->ID );
```
```suggestion
		$this->assertSame( $post_ids[3], $next->ID );
```

:1234: Looks to apply elsewhere so I won't repeat myself.

ramonjd replied:
👍🏻

peterwilsoncc commented on src/wp-includes/link-template.php:
```diff
@@ -1990,6 +1990,11 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 
+	// Only force deterministic fallback if the where clause has not been modified by a filter.
+	if ( $where === $where_prepared ) {
+		$where = $where_prepared_with_deterministic_fallback;
```
Any particular reason you're defining `$where_prepared_with_deterministic_fallback` above rather than here?

ramonjd replied:
I'm trying to reduce my carbon footprint by limiting my Ctrl+V and Ctrl+P  :trollface: 

No reason at all actually. I can move.

ramonjd replied:
Oh I knew there was a better reason than that 😄 

I wanted to use the original `$where` (with post_status conditions) without having to store it in a new var before the filter.

peterwilsoncc commented on src/wp-includes/link-template.php:
```diff
@@ -2004,13 +2009,18 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 	 * @param string  $order   Sort order. 'DESC' for previous post, 'ASC' for next.
 	 */
-	$sort = apply_filters( "get_{$adjacent}_post_sort", "ORDER BY p.post_date $order, p.ID $order LIMIT 1", $post, $order );
+	$sort_prepared = "ORDER BY p.post_date $order LIMIT 1";
```
This will need to be done above the docblock so docs-parsers can figure out the filter it applies to.

peterwilsoncc commented on src/wp-includes/link-template.php:
```diff
@@ -1980,7 +1981,6 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 	 * @since 4.4.0 Added the `$taxonomy` and `$post` parameters.
-	 * @since 6.9.0 Adds ID-based fallback for posts with identical dates in adjacent post queries.
 	 *
 	 * @param string       $where          The `WHERE` clause in the SQL.
```
Yours truly missed this last time.

```suggestion
	 * @param string       $where_prepared The `WHERE` clause in the SQL.
```

peterwilsoncc commented on src/wp-includes/link-template.php:
```diff
@@ -1814,6 +1814,7 @@ function get_next_post( $in_same_term = false, $excluded_terms = '', $taxonomy =
  * Can either be next or previous post.
  *
  * @since 2.5.0
+ * @since 6.9.1 Adds deterministic fallback for sort clause if not modified by a filter.
```
Should probably include the since changes for 6.0.0 that yours truly also missed.

ramonjd replied:
What are they? 😄

peterwilsoncc replied:
Maybe something like this?

```suggestion
 * @since 6.9.0 Introduce deterministic fallback based in IDs to account for date collisions.
 * @since 6.9.1 Remove deterministic fallback for sites modifying the WHERE clause via a filter.
 *                       See https://core.trac.wordpress.org/ticket/64390
```

peterwilsoncc commented on src/wp-includes/link-template.php:
```diff
@@ -2004,13 +2009,18 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 	 * @since 2.5.0
 	 * @since 4.4.0 Added the `$post` parameter.
 	 * @since 4.9.0 Added the `$order` parameter.
-	 * @since 6.9.0 Adds ID sort to ensure deterministic ordering for posts with identical dates.
```
I think it best to keep this and then not that it was removed in 6.9.1 and applied later if changed by a filter.

ramonjd replied:
Ah great thanks, I was wondering about that. Cheers

peterwilsoncc reviewed (commented):
I've added a few notes inline.

ramonjd commented:
Thanks @peterwilsoncc I'll fix those things up next week.

westonruter commented on src/wp-includes/link-template.php:
```diff
@@ -1965,8 +1966,8 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 	// Prepare the where clause for the adjacent post query.
-	$where_prepared = $wpdb->prepare( "WHERE (p.post_date $comparison_operator %s OR (p.post_date = %s AND p.ID $comparison_operator %d)) AND p.post_type = %s $where", $current_post_date, $current_post_date, $post->ID, $post->post_type ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- $comparison_operator is a string literal, either '<' or '>'.
-
+	$where_prepared_with_deterministic_fallback = $wpdb->prepare( "WHERE (p.post_date $comparison_operator %s OR (p.post_date = %s AND p.ID $comparison_operator %d)) AND p.post_type = %s $where", $current_post_date, $current_post_date, $post->ID, $post->post_type ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- $comparison_operator is a string literal, either '<' or '>'.
```
This variable is quite long. How about just `$deterministic_where_prepared`?

ramonjd committed:
Update var name and PHP docs

Unit tests have been updated according to feedback (using assertSame for simple comparisons), and removing filter removes.

ramonjd commented:
Closing as won't fix. See https://core.trac.wordpress.org/ticket/64390#comment:26
